src/bluesky/utils/__init__.py

```diff
@@ -1963,8 +1964,43 @@ def plan(plan):
         Wrapped plans
     """
 
-    @wraps(plan)
+    @wraps_plan(bs_plan, plan)
     def wrapper(*args, **kwargs) -> Plan:
-        return Plan(plan, *args, **kwargs)
+        return Plan(bs_plan, *args, **kwargs)
 
     return wrapper
+
+
+def mark_as_plan(wrapper, wrapped, decorator, **kwargs):
+    """Function that adds attr if plan decorator is applied"""
+
+    wrapper = functools.update_wrapper(wrapper, wrapped, **kwargs)
+    if decorator is plan:
+        wrapper.__is_plan__ = True
+    return wrapper
+
+
+def wraps_plan(wrapped, decorator, **kwargs):
+    """Modified wraps that marks decorated function as bs plan"""
+
+    return functools.partial(mark_as_plan, wrapped=wrapped, decorator=decorator, **kwargs)
+
+
+def is_plan(bs_plan):
+    """Function that checks if function is a generator, or has been marked as plan
+
+    Parameters
+    ----------
+    bs_plan : Function
+        Typically generator coroutine function / Bluesky plan
+
+    Returns
+    -------
+    boolean
+        True if bs_plan arg is a generator, or the __is_plan__ attribute exists and is True.
+    """
+
+    if inspect.isgeneratorfunction(bs_plan) or getattr(bs_plan, "__is_plan__", False):
```

Comment:
Could we also do this as 
```suggestion
    if inspect.isgeneratorfunction(bs_plan) or isintsance(bs_plan, Plan):
```

without having to pick up any extra state?

Reply:
I had tried this originally, but it doesn't appear to work:

```
(venv) [jwlodek@dell-rhel8 bluesky]$ ipython
Python 3.11.9 (main, Jun 19 2024, 10:02:06) [GCC 8.5.0 20210514 (Red Hat 8.5.0-22)]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.23.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from bluesky.plan_stubs import mv

In [2]: from bluesky.utils import is_plan

In [3]: is_plan??

Signature: is_plan(bs_plan)
Source:   
def is_plan(bs_plan):
    if inspect.isgeneratorfunction(bs_plan) or isinstance(bs_plan, Plan):
        return True
    else:
        return False
File:      ~/Workspace/bluesky/src/bluesky/utils/__init__.py
Type:      function

In [4]: is_plan(mv)
Out[4]: False
```

Reply:
oh, right 🐑 

The thing returned by the decorator is a normal function, the thing returned by the normal function is `Plan` which than implements the co-routine functionality when iterated.

Reply:
Indeed, if I do `isinstance(mv(), Plan)` it returns true, but then complains that the plan was not iterated, so maybe just keeping track of the state is easier?

Reply:
Yes, I think tagging it with some private state we then look for in the right implementation.